components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,328 @@
+        }
+    }
+
+    /// Returns an `Iter` of start and stop `u32` points of the UnicodeSet
+    pub fn iter(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Returns the cardinality of the UnicodeSet
+    ///
+    ///
+    pub fn size(&self) -> Result<usize, Box<dyn Error>> {
+        if self.inv_list.len() < 2 {
+            return Err("UnicodeSet length < 2".into());
+        }
+        let end_point_sum: u32 = self.iter().skip(1).step_by(2).sum::<u32>();
+        let start_point_sum: u32 = self.iter().step_by(2).sum::<u32>();
+        Ok((end_point_sum - start_point_sum) as usize)
+    }
+
+    /// Returns whether or not the UnicodeSet is empty
+    pub fn is_empty(&self) -> bool {
+        self.inv_list.len() < 2 // unsure if this is appropriate definition of just self.inv_list.is_empty()
+    }
+
+    /// Wrapper for contains conditions closures
+    fn contains<C>(&self, query: &u32, condition: C) -> bool
+    where
+        C: Fn(usize) -> bool,
+    {
+        match self.inv_list.binary_search(query) {
+            Ok(pos) => {
+                if pos % 2 == 0 {
+                    condition(pos)
+                } else {
+                    false
+                }
+            }
+            Err(pos) => {
+                if pos % 2 != 0 && pos < self.inv_list.len() {
+                    condition(pos)
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    /// Checks to see the query is in the UnicodeSet
+    ///
+    /// Runs a binary search in `O(log(n))` where `n` is the number of start and end points
+    /// in the set using `std::vec::Vec` implementation
+    ///
+    /// Example: `contains_point(&10)`
+    pub fn contains_point(&self, query: &u32) -> bool {
+        let condition_closure = |_: usize| -> bool { true };
+        self.contains(query, condition_closure)
+    }
+
+    /// Checks to see if the range is in the UnicodeSet, returns a Result
+    ///
+    /// Runs a binary search in `O(log(n))` where `n` is the number of start and end points
+    /// in the set using `std::vec::Vec` implementation
+    ///
+    /// Only runs the search once on the `start` parameter, while the `end` parameter is checked
+    /// in a single `O(1)` step
+    ///
+    /// Example: `contains_range(&0, &10)`
+    pub fn contains_range(&self, start: &u32, end: &u32) -> Result<bool, Box<dyn Error>> {
+        if start >= end {
+            return Err("Range cannot be out of order".into());
+        }
+        let condition_closure = |pos: usize| -> bool {
+            if end < &self.inv_list[pos + 1] {
+                true
+            } else {
+                false
+            }
+        };
+        return Ok(self.contains(start, condition_closure));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{parse_serial_string, UnicodeSet, BMP_MAX, CODEPOINT_MAX, CODEPOINT_MIN};
+
+    // parse_serial_string
+    #[test]
+    fn test_parse_serial_string() {
```

Comment:
Not sure if your goal was to mirror the ICU4C or ICU4J implementations, but worth noting that there's the standard trait `FromStr`, which seems as if it would be appropriate in this case.

Reply:
I was considering implementing the `From` trait for `UnicodeSet` from a string, but I was unaware of the `FromStr` trait. Do you agree it would be better if I implemented `FromStr` for UnicodeSet, so for L1 users will call `UnicodeSet::from_str` to generate a UnicodeSet from the serialized string, and for L2 users can use `UnicodeSet::new` to generate the empty UnicodeSet in which builder functions may be applied?

Reply:
IIUC, if you implement `FromStr` for some type (say `YourType`), you can do things like this:

```rust
let t: YourType = "blahblah".parse();
``` 

and the compiler will do the correct thing automatically. (Otherwise, as you note, there'd be no reason to have both `from_str` and `new`.)  

One reason why you may consider to use `FromStr` is because the `.parse()` behavior will be readily understood by all readers of your code -- which is: 'this object is constructed from the text `"blahblah"`, but there's some nontrivial parsing going on inside'.